File: codes/go/claude/partial/task_4/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
)

func allocateResources(patients [][]int, resources []map[string]interface{}, responders int, totalTime int) map[string]interface{} {
	type Assignment struct {
		Patient int `json:"patient"`
		Resource int `json:"resource"`
		Time int `json:"time"`
	}

	assignments := []Assignment{}
	survivors := 0
	totalSurvivalProbability := 0.0

	// Dynamic programming table
	dp := make([][][]float64, len(patients))
	for i := range dp {
		dp[i] = make([][]float64, len(resources))
		for j := range dp[i] {
			dp[i][j] = make([]float64, totalTime+1)
			for k := range dp[i][j] {
				dp[i][j][k] = -1
			}
		}
	}

	// Helper function to calculate probability after applying a resource
	calcProb := func(patient, resource, time int) float64 {
		if time >= len(patients[patient]) {
			return 0
		}
		baseProb := float64(patients[patient][time])
		boost := resources[resource]["boost"].(float64)
		return math.Min(100, baseProb + boost)
	}

	// Recursive function to fill dp table
	var fillDP func(patient, resource, time int) float64
	fillDP = func(patient, resource, time int) float64 {
		if patient >= len(patients) || time > totalTime {
			return 0
		}
		if dp[patient][resource][time] != -1 {
			return dp[patient][resource][time]
		}

		// Don't apply this resource
		maxProb := fillDP(patient+1, 0, time)

		// Apply this resource if possible
		if time + int(resources[resource]["time"].(float64)) <= totalTime {
			prob := calcProb(patient, resource, time)
			prob += fillDP(patient+1, 0, time+int(resources[resource]["time"].(float64)))
			maxProb = math.Max(maxProb, prob)
		}

		dp[patient][resource][time] = maxProb
		return maxProb
	}

	// Fill the dp table
	fillDP(0, 0, 0)

	// Backtrack to find assignments
	currentTime := 0
	for patient := 0; patient < len(patients); patient++ {
		bestResource := -1
		bestTime := -1
		maxProb := 0.0

		for resource := 0; resource < len(resources); resource++ {
			for time := currentTime; time <= totalTime-int(resources[resource]["time"].(float64)); time++ {
				prob := calcProb(patient, resource, time)
				if prob > maxProb {
					maxProb = prob
					bestResource = resource
					bestTime = time
				}
			}
		}

		if bestResource != -1 {
			assignments = append(assignments, Assignment{
				Patient:  patient,
				Resource: bestResource,
				Time:     bestTime,
			})
			currentTime = bestTime + int(resources[bestResource]["time"].(float64))
			totalSurvivalProbability += maxProb
			if maxProb > 50 {
				survivors++
			}
		}
	}

	return map[string]interface{}{
		"assignments":               assignments,
		"survivors":                 survivors,
		"total_survival_probability": totalSurvivalProbability,
	}
}

func main() {
	var input map[string]interface{}
	decoder := json.NewDecoder(os.Stdin)
	err := decoder.Decode(&input)
	if err != nil {
		fmt.Println("Error decoding input:", err)
		return
	}

	patients := input["patients"].([]interface{})
	patientsList := make([][]int, len(patients))
	for i, patient := range patients {
		patientProbs := patient.([]interface{})
		patientsList[i] = make([]int, len(patientProbs))
		for j, prob := range patientProbs {
			patientsList[i][j] = int(prob.(float64))
		}
	}

	resources := input["resources"].([]interface{})
	resourcesList := make([]map[string]interface{}, len(resources))
	for i, resource := range resources {
		resourcesList[i] = resource.(map[string]interface{})
	}

	responders := int(input["responders"].(float64))
	totalTime := int(input["total_time"].(float64))

	result := allocateResources(patientsList, resourcesList, responders, totalTime)

	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	err = encoder.Encode(result)
	if err != nil {
		fmt.Println("Error encoding output:", err)
	}
}

File: codes/python/claude/partial/task_4/Solution.py
import json
import sys

def allocate_resources(patients, resources, responders, total_time):
    def calculate_probability(patient, resource, time):
        base_prob = patient[time] if time < len(patient) else 0
        return min(100, base_prob + resource['boost'])

    def dp_allocation():
        n = len(patients)
        m = len(resources)
        dp = [[[0 for _ in range(total_time + 1)] for _ in range(m + 1)] for _ in range(n + 1)]
        assignments = [[[[] for _ in range(total_time + 1)] for _ in range(m + 1)] for _ in range(n + 1)]

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                for t in range(1, total_time + 1):
                    dp[i][j][t] = dp[i - 1][j][t]
                    assignments[i][j][t] = assignments[i - 1][j][t].copy()

                    if t >= resources[j - 1]['time']:
                        prob = calculate_probability(patients[i - 1], resources[j - 1], t - resources[j - 1]['time'])
                        new_prob = dp[i - 1][j - 1][t - resources[j - 1]['time']] + prob
                        if new_prob > dp[i][j][t]:
                            dp[i][j][t] = new_prob
                            assignments[i][j][t] = assignments[i - 1][j - 1][t - resources[j - 1]['time']].copy()
                            assignments[i][j][t].append((i - 1, j - 1, t - resources[j - 1]['time']))

        return dp[n][m][total_time], assignments[n][m][total_time]

    total_prob, optimal_assignments = dp_allocation()

    final_assignments = []
    used_times = set()
    for patient, resource, time in optimal_assignments:
        if len([a for a in final_assignments if a['time'] <= time < a['time'] + resources[a['resource']]['time']]) < responders:
            final_assignments.append({
                'patient': patient,
                'resource': resource,
                'time': time
            })
            used_times.add(time)

    final_probabilities = [0] * len(patients)
    for assignment in final_assignments:
        patient = assignment['patient']
        resource = assignment['resource']
        time = assignment['time']
        final_probabilities[patient] = calculate_probability(patients[patient], resources[resource], time)

    survivors = sum(1 for prob in final_probabilities if prob > 50)
    total_survival_probability = sum(final_probabilities)

    return {
        'assignments': final_assignments,
        'survivors': survivors,
        'total_survival_probability': total_survival_probability
    }

def main():
    input_data = json.loads(sys.stdin.read())
    result = allocate_resources(
        input_data['patients'],
        input_data['resources'],
        input_data['responders'],
        input_data['total_time']
    )
    print(json.dumps(result))

if __name__ == "__main__":
    main()